### PR TITLE
Upstream mutual TLS handshake support

### PR DESCRIPTION
**Pull request by dndx**
This is achieved by specifying the Certificate to be used during TLS
handshake via the newly added `client_certificate` property on the Service
core entity.

Note: Based on https://github.com/Kong/kong/pull/4797, review that PR first.

### CODE REVIEW

File: kong/pdk/service/request.lua [hunk 1]
```diff
@@ -417,6 +419,49 @@ local function new(self)
   end
 
 
+  if ngx.config.subsystem == "http" then
```
Comment (james-callahan) [inline kong/pdk/service/request.lua]: Why not support this in stream mode?
Should we provide a stub/error function in stream mode until if we are delaying the work?

File: kong/runloop/certificate.lua [hunk 2]
```diff
@@ -119,15 +119,23 @@ local function fetch_sni(sni, i)
 end
 
 
-local function fetch_certificate(sni)
-  local certificate, err = singletons.db.certificates:select(sni.certificate)
+local function fetch_certificate(pkey, sni_name)
```
Comment (james-callahan) [inline kong/runloop/certificate.lua]: I was confused by this variable name: `pkey` is frequently used to mean "private key" in certificate handling code.

File: kong/db/schema/entities/cas.lua [hunk 3]
```diff
@@ -0,0 +1,41 @@
+local typedefs = require "kong.db.schema.typedefs"
+local openssl_x509 = require "openssl.x509"
+local ngx_time = ngx.time
```
Comment (thibaultcha) [inline kong/db/schema/entities/cas.lua]: Note that we don't have to cache those global lookups in the Admin API server, it being a very cold path compared to our proxy server.

File: kong/db/schema/entities/cas.lua [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+return {
+  name        = "cas",
+  primary_key = { "id" },
+
+  fields = {
+    { id             = typedefs.uuid, },
+    { created_at     = typedefs.auto_timestamp_s },
+    { cert           = typedefs.certificate { required = true }, },
+    { tags           = typedefs.tags },
```
Comment (thibaultcha) [inline kong/db/schema/entities/cas.lua]: style: we find it to be easier not to align `=` signs anymore, if you care to revert the change... Not a blocker for me.

File: kong/db/schema/entities/cas.lua [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+          return nil, "certificate unusable, \"Not Before\" time is " ..
+                      "in the future"
+        end
+
+        if not_after < now then
+          return nil, "certificate expired, \"Not After\" time is in the past"
+        end
+
+        if not cert:getBasicConstraints("CA") then
+          return nil, "certificate does not have \"CA\" basic constrain set"
```
Comment (thibaultcha) [inline kong/db/schema/entities/cas.lua]: Should we produce a more "non SSL expert" error message? Something like `"certificate does not appear to be a CA (missing "CA" basic constraint)"`.

Regardless, we should fix the typo too ("constraint").

File: kong/constants.lua [hunk 6]
```diff
@@ -87,9 +87,9 @@ return {
   -- schemas of dependencies need to be loaded first.
   CORE_ENTITIES = {
     "consumers",
+    "certificates",
```
[the remark this answers is not in the document]
Comment (thibaultcha) [reply kong/constants.lua]: This list is not supposed to be alphabetically sorted (not should it give the impression that it is) since its order is semantically meaningful?

File: kong/runloop/certificate.lua (hunk 2 above)
Comment (thibaultcha) [reply kong/runloop/certificate.lua]: Agreed, we use `pk` or something similar in our DAO code when we mean "primary key". We should not use `pkey` here (I too got confused)

File: kong/runloop/certificate.lua [hunk 7]
```diff
@@ -119,15 +119,23 @@ local function fetch_sni(sni, i)
...
+  local certificate, err = singletons.db.certificates:select(pkey)
   if err then
-    return nil, "failed to fetch certificate for '" .. sni.name .. "' SNI: " ..
-                err
+    if sni_name then
+      return nil, "failed to fetch certificate for '" .. sni_name .. "' SNI: " ..
+                  err
+    end
+
+    return "failed to fetch certificate " .. pkey.id
```
Comment (thibaultcha) [inline kong/runloop/certificate.lua]: A big mistake here I believe: missing `return nil` (first return value)? This would make the above string the valid, cached return value of `fetch_certificate`

File: kong/db/schema/entities/cas.lua (hunk 4 above)
Comment (dndx) [reply kong/db/schema/entities/cas.lua]: This has been reverted.

File: kong/db/schema/entities/cas.lua (hunk 5 above)
Comment (dndx) [reply kong/db/schema/entities/cas.lua]: Changed to `certificate does not appear to belong to a CA (missing the "CA" basic constraint)`.

File: kong/constants.lua (hunk 6 above)
[the remark this answers is not in the document]
Comment (dndx) [reply kong/constants.lua]: Because `service` has FK reference to `certificates`, it has to load before `service` entity schema is loaded.

[the remark this answers is not in the document]
Comment (thibaultcha) [reply kong/constants.lua]: Yep, that was clear. Was just not agreeing with @james-callahan on this one that it had to be alphabetically sorted.

File: kong/runloop/certificate.lua (hunk 2 above)
Comment (dndx) [reply kong/runloop/certificate.lua]: Fixed. Thanks for pointing out!

File: kong/runloop/certificate.lua (hunk 7 above)
Comment (dndx) [reply kong/runloop/certificate.lua]: Good catch @thibaultcha, that was fixed!

File: kong/pdk/service/request.lua (hunk 1 above)
Comment (dndx) [reply kong/pdk/service/request.lua]: Talked with @thibaultcha, it should be alright to skip this for now until someone complaints (and it is not a silent failure)...

**Comment by dndx** [comment]
All comments addressed. Also pushed a new commit https://github.com/Kong/kong/pull/4800/commits/d1c8121be0404d8da69c010aaa369a217bd40042 that adds additional schema constraints for the new field. Please take a peek @thibaultcha @hishamhm :).

File: kong/db/migrations/core/005_120_to_130.lua [hunk 8]
```diff
@@ -55,9 +68,10 @@ return {
         PRIMARY KEY (partition, id)
       );
 
-
-
       ALTER TABLE routes ADD headers map<text,frozen<set<text>>>;
+      ALTER TABLE services ADD client_certificate_id uuid;
+
+      CREATE INDEX IF NOT EXISTS services_client_certificate_id_idx ON services(client_certificate_id);
```
Comment (thibaultcha) [inline kong/db/migrations/core/005_120_to_130.lua]: style: in terms of migrations styling, we follow a rule to have 3 line breaks between statements related to a given feature (hence why the line breaks you removed here. We apply this for both Postgres and Cassandra migrations.

File: spec/02-integration/05-proxy/18-client_certificate_spec.lua [hunk 9]
```diff
@@ -0,0 +1,135 @@
+local helpers = require "spec.helpers"
+local ssl_fixtures = require "spec.fixtures.ssl"
+
+
+local fixtures = {
+  http_mock = {
+    upstream_mtls = [[
+      server {
+          server_name example.com;
+          listen 16798 ssl;
```
Comment (thibaultcha) [inline spec/02-integration/05-proxy/18-client_certificate_spec.lua]: If possible, it'd be better to use a unix socket here, but not a blocker.

**Comment by thibaultcha** [comment]
Agreed with @kikito on the git history cleanup regarding fix commits as well

File: spec/02-integration/05-proxy/18-client_certificate_spec.lua (hunk 9 above)
Comment (dndx) [reply spec/02-integration/05-proxy/18-client_certificate_spec.lua]: I don't think we create Services with unix domain sockets as upstream in the test suite so most likely it is not supported.
